app/models/manageiq/providers/azure/cloud_manager/metrics_capture.rb

```diff
@@ -77,76 +77,135 @@ def perf_collect_metrics(interval_name, start_time = nil, end_time = nil)
       metrics = metrics.map do |metric|
-        [metric.name.value, metric.timeseries.flat_map do |t|
+        metric_values = metric.timeseries.flat_map do |t|
           t.data.select { |d| d.respond_to?(:average) }
-        end]
-      end.to_h.freeze
+        end
+        timestamped_values = metric_values.each_with_object({}) do |metric_value, memo|
+          timestamp = parse_timestamp(metric_value.time_stamp)
+          (memo[timestamp] ||= []) << metric_value.average
+        end
+        [metric.name.value, timestamped_values]
+      end.to_h
 
-      counter_values = api_counters_info.each_with_object({}) do |counter_info, memo|
-        counter_metrics = metrics.values_at(*counter_info.metrics).compact.flatten
+      ## Counters, which are available only through legacy API and storage account (raw)
 
-        # { timestamp => [value, ...], ... }
-        timestamped_values = counter_metrics.each_with_object({}) do |metric_value, ts_memo|
-          timestamp = Time.zone.parse(metric_value.time_stamp)
-          (ts_memo[timestamp] ||= []) << metric_value.average
+      storage_conn = Azure::Armrest::StorageAccountService.new(@connection)
+      storage_accounts = storage_conn.list_all
+
+      metric_definitions, _timings = Benchmark.realtime_block(:capture_counters) do
+        # azure-armrest gem needs to be modified to accept version ('list_definitions' method)
+        response = metrics_conn.send(:rest_get, definitions_url.to_s)
+        Azure::Armrest::ArmrestCollection.create_from_response(response, Azure::Armrest::Insights::MetricDefinition)
+      end
+
+      metric_definitions.each do |metric_definition|
+        metric_name = metric_definition.name.value
+        next unless METRIC_NAMES[:raw].include?(metric_name)
+
+        metric_availability = metric_definition.metric_availabilities.detect { |ma| ma.time_grain == INTERVAL_1_MINUTE }
+        next unless metric_availability
+
+        timestamped_values = metrics[metric_name] ||= {}
+
+        metric_location = metric_availability.location
+
+        storage_account_name = URI.parse(metric_location.table_endpoint).host.split('.').first
+        storage_account      = storage_accounts.find { |account| account.name == storage_account_name }
+        storage_account_keys = storage_conn.list_account_keys(storage_account.name, storage_account.resource_group)
+        storage_account_key  = storage_account_keys.fetch('key1')
+
+        metric_location.table_info.each do |table_info|
+          t_start_time = parse_timestamp(table_info.start_time)
+          t_end_time   = parse_timestamp(table_info.end_time).end_of_day
+          next unless time_interval.overlaps?(t_start_time..t_end_time)
+
+          filter = [
+            "CounterName eq '#{metric_name}'",
+            "PartitionKey eq '#{metric_location.partition_key}'",
+            "Timestamp ge datetime'#{@start_time}'",
+            "Timestamp le datetime'#{@end_time}'",
+          ].join(' and ')
```

Comment:
Much benefit of joining an array of essentially constant strings with some interpolation vs just one long string?  Did you break it up like this to avoid 125 char line length limit?

Reply:
I did that for readability reasons. It is much easier to say we have 4 conditions here, that when reading a single line.

Reply:
You could make it a multiline string with a heredoc and not incur the runtime overhead of an array join (minimal sure, but unnecessary)

Reply:
are we getting into the tastes discussion again? you like your approach, I like mine, and this question price is such a nothing. how big should be performance win you proposing?

Reply:
From the POV of a reviewer (all opinions my own) there are three types of taste discussion:

1. Personal Preference for which there is no significant upside or downside

An example of this might be if you really like using `class << self` vs `def self.` or you like `%i[a b]` instead of `[:a, :b]`

In this case I couldn't care less which you choose and really try to keep my personal preferences out of the review unless as a project we've standardized on one (e.g. %i[] instead of %i() was a deliberate change to the rubocop rules at the project level).

2. Keeping consistent with the current code vs refactoring

In this case while a different design might be objectively better, usually it is preferable to stay consistent with the rest of the method/file/codebase.

I believe this is the issue with making `@connection`, `@start_time`, and `@end_time` instance vars when they are currently intentionally method parameters.

Would it be better if the `MetricsCapture` class took `@start_time` and `@end_time` as instance params, probably.  And if we want to do that we should come up with the design in core and apply it to the providers and properly refactor it.

3. Cases where there don't appear to be any upsides with some downsides

That is what I believe this case is.  There is literally no readability improvement of this over
```
<<~QUERY
  CounterName eq '#{metric_name}' AND \
  PartitionKey eq '#{metric_location.partition_key}' AND \
  Timestamp ge datetime'#{@start_time}' AND \
  Timestamp le datetime'#{@end_time}'
QUERY
```

and what you have here might be your taste but is objectively worse (even if only slightly) with marginal to no benefits.